cmd/launcher/watchdog/watchdog_service_windows.go

```diff
@@ -0,0 +1,231 @@
+			w.systemSlogger.Log(ctx, slog.LevelInfo,
+				"runLauncherWatchdogService exited",
+				"err", err,
+				"stack_trace", fmt.Sprintf("%+v", errors.WithStack(err)),
+			)
+		} else {
+			w.systemSlogger.Log(ctx, slog.LevelInfo,
+				"runLauncher exited cleanly",
+			)
+		}
+
+		// signal to fully exit so that the service manager can restart the service
+		runRestartServiceResults <- struct{}{}
+	}, func(r any) {
+		w.systemSlogger.Log(ctx, slog.LevelError,
+			"exiting after runLauncherWatchdogService panic",
+			"err", r,
+		)
+
+		// signal to fully exit so that the service manager can restart the service.
+		runRestartServiceResults <- struct{}{}
+	})
+
+	for {
+		select {
+		case c := <-r:
+			switch c.Cmd {
+			case svc.Interrogate:
+				changes <- c.CurrentStatus
+				// Testing deadlock from https://code.google.com/p/winsvc/issues/detail?id=4
+				time.Sleep(100 * time.Millisecond)
+				changes <- c.CurrentStatus
+			case svc.Stop, svc.Shutdown:
+				w.systemSlogger.Log(ctx, slog.LevelInfo, "shutdown request received")
+				changes <- svc.Status{State: svc.StopPending}
+				cancel()
+				time.Sleep(1 * time.Second) // give checker routine enough time to shut down
+				changes <- svc.Status{State: svc.Stopped, Accepts: cmdsAccepted}
+				return ssec, errno
+			default:
+				w.systemSlogger.Log(ctx, slog.LevelInfo,
+					"unexpected change request",
+					"service", launcherWatchdogServiceName,
+					"change_request", fmt.Sprintf("%+v", c),
+				)
+			}
+		case <-runRestartServiceResults:
+			w.systemSlogger.Log(ctx, slog.LevelInfo,
+				"shutting down restart service after exit",
+			)
+			// We don't want to tell the service manager that we've stopped on purpose,
+			// so that the service manager will restart launcher correctly.
+			// We use this error code largely because the windows/svc code also uses it
+			// and it seems semantically correct enough; it doesn't appear to matter to us
+			// what the code is.
+			return false, uint32(windows.ERROR_EXCEPTION_IN_SERVICE)
+		}
+	}
+}
+
+func (w *winWatchdogSvc) checkLauncherStatus(ctx context.Context) error {
+	serviceManager, err := mgr.Connect()
+	if err != nil {
+		w.slogger.Log(ctx, slog.LevelError,
+			"connecting to service control manager",
+			"err", err,
+		)
+
+		return err
+	}
+
+	defer serviceManager.Disconnect()
+
+	launcherService, err := serviceManager.OpenService(launcherServiceName)
+	if err != nil {
+		return fmt.Errorf("opening launcher service: %w", err)
+	}
+
+	defer launcherService.Close()
+
+	currentStatus, err := launcherService.Query()
+	if err != nil {
+		return fmt.Errorf("checking current launcher status: %w", err)
+	}
+
+	if currentStatus.State == svc.Stopped {
+		w.slogger.Log(ctx, slog.LevelInfo, "watchdog service checker detected stopped state, restarting")
+		return launcherService.Start()
+	}
+
+	return nil
+}
+
+func runLauncherWatchdogService(ctx context.Context, w *winWatchdogSvc) error {
+	ticker := time.NewTicker(1 * time.Minute)
```

Comment:
I wonder if this frequency will be okay. Or if there's going to be some memory leak. Well, it's behind a feature flag, so we'll get to find out!